### PR TITLE
 Add file to JSON source info

### PR DESCRIPTION
**Pull request by yufeih**
Populate `file` to `SourceInfo` during json parse. Currently only line/column is passed to `JToken`.

### CODE REVIEW

**Comment by OsmondJiang** [comment]
what are we going to resolve in this PR?

File: src/docfx/lib/json/SourceInfoJsonConverter.cs [hunk 1]
```diff
@@ -15,14 +16,13 @@ public override bool CanConvert(Type objectType)
...
-            var source = JsonUtility.ToSourceInfo((IJsonLineInfo)reader);
+            var source = reader is JTokenReader tokenReader ? JsonUtility.GetSourceInfo(tokenReader.CurrentToken) : null;
             var valueType = objectType.GenericTypeArguments[0];
             var value = serializer.Deserialize(reader, valueType);
 
             if (value is null)
                 return null;
 
-            // TODO: populate file info
             return Activator.CreateInstance(objectType, value, source);
```
Comment (yufeih) [inline src/docfx/lib/json/SourceInfoJsonConverter.cs]: `file` is populated here, each `JToken` is a associated with a different source info.

File: src/docfx/lib/json/JsonUtility.cs [hunk 2]
```diff
@@ -220,13 +220,11 @@ public static (List<Error>, T) ToObject<T>(JToken token)
         /// Parse a string to JToken.
         /// Validate null value during the process.
         /// </summary>
-        // TODO: Pass in file to be filled into SourceInfo
-        public static (List<Error>, JToken) Parse(string json)
+        public static (List<Error>, JToken) Parse(string json, string file)
```
Comment (yufeih) [inline src/docfx/lib/json/JsonUtility.cs]: `file` is originally passed here.

**Comment by live1206** [comment]
>             File = file;

Should we add a not null check here? so that there is always a File populated.
And populate file for these as well?
[https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L30](https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L30)
[https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L46](https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L46)

---
Refers to: src/docfx/lib/log/SourceInfo.cs:37 in 1e80a45. [](commit_id = 1e80a45e458b9ac292e037a56b719cd28d62af20, deletion_comment = False)

**Comment by yufeih** [comment]
> > ```
> >         File = file;
> > ```
> 
> Should we add a not null check here? so that there is always a File populated.
> And populate file for these as well?
> [docfx/src/docfx/build/metadata/MetadataValidator.cs](https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L30)
> 
> Line 30 in [aefc5c9](/dotnet/docfx/commit/aefc5c92856e969675de894a49255f44197a991d)
> 
>  errors.Add(Errors.ReservedMetadata(new SourceInfo(null, lineInfo?.LineNumber ?? 0, lineInfo?.LinePosition ?? 0), key, token.Path)); 
> 
> [docfx/src/docfx/build/metadata/MetadataValidator.cs](https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L46)
> 
> Line 46 in [aefc5c9](/dotnet/docfx/commit/aefc5c92856e969675de894a49255f44197a991d)
> 
>  new SourceInfo(null, nestedLineInfo?.LineNumber ?? 0, nestedLineInfo?.LinePosition ?? 0), 
> Refers to: src/docfx/lib/log/SourceInfo.cs:37 in [1e80a45](https://github.com/dotnet/docfx/commit/1e80a45e458b9ac292e037a56b719cd28d62af20). [](commit_id = [1e80a45](https://github.com/dotnet/docfx/commit/1e80a45e458b9ac292e037a56b719cd28d62af20), deletion_comment = False)

yes, after all the other `nulls` are cleaned up.

**Comment by live1206** [comment]
> > > ```
> > >         File = file;
> > > ```
> > 
> > 
> > Should we add a not null check here? so that there is always a File populated.
> > And populate file for these as well?
> > [docfx/src/docfx/build/metadata/MetadataValidator.cs](https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L30)
> > Line 30 in [aefc5c9](/dotnet/docfx/commit/aefc5c92856e969675de894a49255f44197a991d)
> > errors.Add(Errors.ReservedMetadata(new SourceInfo(null, lineInfo?.LineNumber ?? 0, lineInfo?.LinePosition ?? 0), key, token.Path));
> > [docfx/src/docfx/build/metadata/MetadataValidator.cs](https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L46)
> > Line 46 in [aefc5c9](/dotnet/docfx/commit/aefc5c92856e969675de894a49255f44197a991d)
> > new SourceInfo(null, nestedLineInfo?.LineNumber ?? 0, nestedLineInfo?.LinePosition ?? 0),
> > Refers to: src/docfx/lib/log/SourceInfo.cs:37 in [1e80a45](https://github.com/dotnet/docfx/commit/1e80a45e458b9ac292e037a56b719cd28d62af20). [](commit_id = [1e80a45](https://github.com/dotnet/docfx/commit/1e80a45e458b9ac292e037a56b719cd28d62af20), deletion_comment = False)
> 
> yes, after all the other `nulls` are cleaned up.

As I can see, the two in `MetadataValidator.cs` are the only nulls left to clean up.

**Comment by yufeih** [comment]
> > > > ```
> > > >         File = file;
> > > > ```
> > > 
> > > 
> > > Should we add a not null check here? so that there is always a File populated.
> > > And populate file for these as well?
> > > [docfx/src/docfx/build/metadata/MetadataValidator.cs](https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L30)
> > > Line 30 in [aefc5c9](/dotnet/docfx/commit/aefc5c92856e969675de894a49255f44197a991d)
> > > errors.Add(Errors.ReservedMetadata(new SourceInfo(null, lineInfo?.LineNumber ?? 0, lineInfo?.LinePosition ?? 0), key, token.Path));
> > > [docfx/src/docfx/build/metadata/MetadataValidator.cs](https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L46)
> > > Line 46 in [aefc5c9](/dotnet/docfx/commit/aefc5c92856e969675de894a49255f44197a991d)
> > > new SourceInfo(null, nestedLineInfo?.LineNumber ?? 0, nestedLineInfo?.LinePosition ?? 0),
> > > Refers to: src/docfx/lib/log/SourceInfo.cs:37 in [1e80a45](https://github.com/dotnet/docfx/commit/1e80a45e458b9ac292e037a56b719cd28d62af20). [](commit_id = [1e80a45](https://github.com/dotnet/docfx/commit/1e80a45e458b9ac292e037a56b719cd28d62af20), deletion_comment = False)
> > 
> > 
> > yes, after all the other `nulls` are cleaned up.
> 
> As I can see, the two in `MetadataValidator.cs` are the only nulls left to clean up.

https://github.com/dotnet/docfx/commit/17b9711e9305d9d3bea68eeb7a5186ca5865b4c1#diff-600d735a03558a7e04fbe7a52cbea80aR20 makes that line could potentially also be `null`

**Comment by live1206** [comment]
> > > > > ```
> > > > >         File = file;
> > > > > ```
> > > > 
> > > > 
> > > > Should we add a not null check here? so that there is always a File populated.
> > > > And populate file for these as well?
> > > > [docfx/src/docfx/build/metadata/MetadataValidator.cs](https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L30)
> > > > Line 30 in [aefc5c9](/dotnet/docfx/commit/aefc5c92856e969675de894a49255f44197a991d)
> > > > errors.Add(Errors.ReservedMetadata(new SourceInfo(null, lineInfo?.LineNumber ?? 0, lineInfo?.LinePosition ?? 0), key, token.Path));
> > > > [docfx/src/docfx/build/metadata/MetadataValidator.cs](https://github.com/dotnet/docfx/blob/aefc5c92856e969675de894a49255f44197a991d/src/docfx/build/metadata/MetadataValidator.cs#L46)
> > > > Line 46 in [aefc5c9](/dotnet/docfx/commit/aefc5c92856e969675de894a49255f44197a991d)
> > > > new SourceInfo(null, nestedLineInfo?.LineNumber ?? 0, nestedLineInfo?.LinePosition ?? 0),
> > > > Refers to: src/docfx/lib/log/SourceInfo.cs:37 in [1e80a45](https://github.com/dotnet/docfx/commit/1e80a45e458b9ac292e037a56b719cd28d62af20). [](commit_id = [1e80a45](https://github.com/dotnet/docfx/commit/1e80a45e458b9ac292e037a56b719cd28d62af20), deletion_comment = False)
> > > 
> > > 
> > > yes, after all the other `nulls` are cleaned up.
> > 
> > 
> > As I can see, the two in `MetadataValidator.cs` are the only nulls left to clean up.
> 
> [17b9711#diff-600d735a03558a7e04fbe7a52cbea80aR20](https://github.com/dotnet/docfx/commit/17b9711e9305d9d3bea68eeb7a5186ca5865b4c1#diff-600d735a03558a7e04fbe7a52cbea80aR20) makes that line could potentially also be `null`

ok, leave it for later change then

**Pull request merged**